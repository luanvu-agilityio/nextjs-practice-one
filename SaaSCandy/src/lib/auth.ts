'use server';

import { redirect } from 'next/navigation';
import { authApi } from '@/api';
import { ROUTES } from '@/constants';
import type { AuthState } from '@/types';

export async function signInAction(formData: FormData): Promise<AuthState> {
  const email = formData.get('email') as string;
  const password = formData.get('password') as string;

  try {
    // Validate with MockAPI
    const response = await authApi.login({ email, password });

    if (!response.user) {
      return {
        error: 'Invalid email or password',
        success: undefined,
        user: undefined,
        credentials: undefined,
      };
    }

    return {
      success: true,
      user: response.user,
      credentials: { email, password },
      error: undefined,
    };
  } catch (error) {
    console.error('Sign in error:', error);
    return {
      error: error instanceof Error ? error.message : 'Sign in failed',
      success: undefined,
      user: undefined,
      credentials: undefined,
    };
  }
}

export async function signUpAction(formData: FormData): Promise<AuthState> {
  const name = formData.get('name') as string;
  const email = formData.get('email') as string;
  const password = formData.get('password') as string;

  try {
    const [firstName, lastName] = name.split(' ');

    const response = await authApi.register({
      email,
      password,
      firstName: firstName || name,
      lastName: lastName || '',
    });

    if (!response.user) {
      return {
        error: 'Registration failed',
        success: undefined,
        user: undefined,
        credentials: undefined,
      };
    }

    return {
      success: true,
      user: response.user,
      credentials: { email, password },
      error: undefined,
    };
  } catch (error) {
    console.error('Sign up error:', error);
    return {
      error: error instanceof Error ? error.message : 'Sign up failed',
      success: undefined,
      user: undefined,
      credentials: undefined,
    };
  }
}

export async function signOutAction() {
  redirect(ROUTES.HOME);
}
