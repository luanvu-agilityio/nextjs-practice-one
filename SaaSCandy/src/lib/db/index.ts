import { drizzle } from 'drizzle-orm/postgres-js';
import postgres from 'postgres';
import * as schema from './schema';

const connectionString = process.env.POSTGRES_URL!;
export const db = drizzle(postgres(connectionString), { schema });
export * from './schema';
