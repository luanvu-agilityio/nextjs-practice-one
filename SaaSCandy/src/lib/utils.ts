/**
 * Utility function for merging Tailwind CSS class names conditionally.
 *
 * - Uses `clsx` for conditional class logic.
 * - Uses `tailwind-merge` to resolve Tailwind class conflicts.
 *
 * @param inputs - List of class values (strings, arrays, objects).
 * @returns A single merged class name string.

 */
import { clsx, type ClassValue } from 'clsx';
import { twMerge } from 'tailwind-merge';

export function cn(...inputs: ClassValue[]) {
  return twMerge(clsx(inputs));
}
