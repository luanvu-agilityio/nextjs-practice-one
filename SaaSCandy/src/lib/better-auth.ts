import { betterAuth } from 'better-auth';
import { drizzleAdapter } from 'better-auth/adapters/drizzle';
import { twoFactor } from 'better-auth/plugins';
import { Resend } from 'resend';
import { db } from './db';
import {
  ResetPasswordEmail,
  VerificationEmail,
} from '@/constants/email-template';

const resend = new Resend(process.env.RESEND_API_KEY);

export const auth = betterAuth({
  database: drizzleAdapter(db, {
    provider: 'pg',
  }),

  secret: process.env.BETTER_AUTH_SECRET!,
  baseURL: process.env.BETTER_AUTH_URL || 'http://localhost:3000',

  emailVerification: {
    enabled: true,
    sendOnSignUp: true,
    autoSignInAfterVerification: true,
    requireEmailVerification: true,
    sendVerificationEmail: async ({ user, token }) => {
      const verificationUrl = `${process.env.BETTER_AUTH_URL}/email-verification?token=${token}`;

      await resend.emails.send({
        from: process.env.RESEND_FROM_EMAIL || 'onboarding@resend.dev',
        to: user.email,
        subject: 'Verify your email - SaaSCandy',
        html: VerificationEmail({ verificationUrl }),
      });
    },
  },

  emailAndPassword: {
    enabled: true,
    requireEmailVerification: true,
    sendResetPassword: async ({ user, token }) => {
      const resetUrl = `${process.env.BETTER_AUTH_URL}/reset-password?token=${token}`;

      await resend.emails.send({
        from: process.env.RESEND_FROM_EMAIL || 'onboarding@resend.dev',
        to: user.email,
        subject: 'Reset your password - SaaSCandy',
        html: ResetPasswordEmail({ resetUrl }),
      });
    },
  },

  plugins: [
    twoFactor({
      issuer: 'SaaSCandy',
      otpOptions: {
        period: 30,
        digits: 6,
      },
    }),
  ],

  socialProviders: {
    google: {
      clientId: process.env.GOOGLE_CLIENT_ID || '',
      clientSecret: process.env.GOOGLE_CLIENT_SECRET || '',
      enabled: !!process.env.GOOGLE_CLIENT_ID,
    },
    github: {
      clientId: process.env.GITHUB_ID || '',
      clientSecret: process.env.GITHUB_SECRET || '',
      enabled: !!process.env.GITHUB_ID,
    },
  },

  logger: {
    level: 'debug',
  },
});

export type Session = typeof auth.$Infer.Session;
export type User = (typeof auth.$Infer.Session)['user'];
