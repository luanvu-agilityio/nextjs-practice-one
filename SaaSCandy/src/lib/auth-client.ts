/**
 * Client-side authentication utilities using Better Auth for React.
 *
 * - Provides sign-in, sign-up, sign-out, session hooks, and user update functions.
 * - Supports two-factor authentication (2FA) via the twoFactorClient plugin.
 * - Uses the app's base URL from environment variables.
 *
 * Usage:
 *   import { signIn, useSession } from '@/lib/auth-client';
 */
import { createAuthClient } from 'better-auth/react';
import { twoFactorClient } from 'better-auth/client/plugins';

export const authClient = createAuthClient({
  baseURL: process.env.NEXT_PUBLIC_APP_URL || 'http://localhost:3000',
  plugins: [twoFactorClient()],
});

export const {
  signIn,
  signUp,
  signOut,
  useSession,
  getSession,
  updateUser,
  twoFactor: { enable: enableTwoFactor, disable: disableTwoFactor, verifyOtp },
} = authClient;
