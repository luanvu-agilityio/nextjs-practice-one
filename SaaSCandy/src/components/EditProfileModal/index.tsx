'use client';

import { useState } from 'react';
import { useForm } from 'react-hook-form';
import { zodResolver } from '@hookform/resolvers/zod';
import { useSession } from '@/lib/auth-client';

// Components
import { Button } from '@/components/common';
import { InputController } from '@/components/common/InputController';
import {
  Dialog,
  DialogContent,
  DialogFooter,
  DialogHeader,
  DialogTitle,
} from '@/components';
import { ErrorMessage } from '../common/ErrorMessage';

// Utils
import { EditProfileFormData, editProfileSchema } from '@/utils';

// Service
import { updateProfile } from '@/service';

interface EditProfileModalProps {
  open: boolean;
  onOpenChange: (open: boolean) => void;
  onSuccess: () => void;
}

function EditProfileModal({
  open,
  onOpenChange,
  onSuccess,
}: Readonly<EditProfileModalProps>) {
  const { data: session } = useSession();
  const [isSubmitting, setIsSubmitting] = useState(false);
  const [error, setError] = useState<string | null>(null);

  const user = session?.user;

  const { control, handleSubmit, reset } = useForm<EditProfileFormData>({
    resolver: zodResolver(editProfileSchema),
    mode: 'onBlur',
    defaultValues: {
      firstName: user?.name?.split(' ')[0] || '',
      lastName: user?.name?.split(' ')[1] || '',
      email: user?.email || '',
    },
  });

  const onSubmit = async (data: EditProfileFormData) => {
    if (!user?.id) {
      setError('User session not found');
      return;
    }

    setIsSubmitting(true);
    setError(null);

    try {
      const fullName = `${data.firstName || ''} ${data.lastName || ''}`.trim();

      const result = await updateProfile({
        name: fullName,
        email: data.email,
      });

      if (!result.success) {
        throw new Error(result.error || 'Failed to update profile');
      }

      // Reload to refresh session
      window.location.reload();

      setTimeout(() => {
        onSuccess();
        onOpenChange(false);
        reset();
      }, 100);
    } catch (error) {
      setError(
        error instanceof Error ? error.message : 'Failed to update profile'
      );
      setIsSubmitting(false);
    }
  };

  const handleClose = () => {
    onOpenChange(false);
    reset();
    setError(null);
  };

  return (
    <Dialog open={open} onOpenChange={handleClose}>
      <DialogContent className='sm:max-w-[425px]'>
        <DialogHeader>
          <DialogTitle>Edit Profile</DialogTitle>
        </DialogHeader>

        <form onSubmit={handleSubmit(onSubmit)} className='space-y-4'>
          {error && <ErrorMessage customMessage={error} />}

          <div className='grid grid-cols-2 gap-4'>
            <InputController
              name='firstName'
              label='First Name'
              control={control}
              placeholder='Enter first name'
              disabled={isSubmitting}
            />
            <InputController
              name='lastName'
              label='Last Name'
              control={control}
              placeholder='Enter last name'
              disabled={isSubmitting}
            />
          </div>

          <InputController
            name='email'
            label='Email Address'
            control={control}
            type='email'
            placeholder='Enter email address'
            disabled={isSubmitting}
          />

          <DialogFooter>
            <Button
              type='button'
              variant='secondary'
              onClick={handleClose}
              disabled={isSubmitting}
            >
              Cancel
            </Button>
            <Button type='submit' variant='primary' disabled={isSubmitting}>
              {isSubmitting ? 'Saving...' : 'Save Changes'}
            </Button>
          </DialogFooter>
        </form>
      </DialogContent>
    </Dialog>
  );
}

export default EditProfileModal;
