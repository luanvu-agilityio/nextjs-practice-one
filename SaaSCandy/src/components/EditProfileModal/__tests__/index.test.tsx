import { render, screen } from '@testing-library/react';
import EditProfileModal from '@/components/EditProfileModal';
import { useSession } from '@/lib/auth-client';

jest.mock('@/lib/auth-client');
jest.mock('@/service');

const mockUseSession = useSession as jest.MockedFunction<typeof useSession>;

describe('EditProfileModal - Snapshot Tests', () => {
  beforeEach(() => {
    mockUseSession.mockReturnValue({
      data: {
        user: {
          id: '123',
          name: 'John Doe',
          email: 'john@example.com',
          createdAt: new Date(),
          updatedAt: new Date(),
          emailVerified: false,
          twoFactorEnabled: undefined,
          image: null,
        },
        session: {
          id: '',
          createdAt: new Date(),
          updatedAt: new Date(),
          userId: '',
          expiresAt: new Date(),
          token: '',
        },
      },
      isPending: false,
      error: null,
      refetch: jest.fn(),
    });
  });

  it('should match snapshot when open', () => {
    const { container } = render(
      <EditProfileModal
        open={true}
        onOpenChange={() => {}}
        onSuccess={() => {}}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it('should populate form with user data', () => {
    render(
      <EditProfileModal
        open={true}
        onOpenChange={() => {}}
        onSuccess={() => {}}
      />
    );

    const firstNameInput = screen.getByPlaceholderText(
      'Enter first name'
    ) as HTMLInputElement;
    const lastNameInput = screen.getByPlaceholderText(
      'Enter last name'
    ) as HTMLInputElement;
    const emailInput = screen.getByPlaceholderText(
      'Enter email address'
    ) as HTMLInputElement;

    expect(firstNameInput.value).toBe('John');
    expect(lastNameInput.value).toBe('Doe');
    expect(emailInput.value).toBe('john@example.com');
  });
});
