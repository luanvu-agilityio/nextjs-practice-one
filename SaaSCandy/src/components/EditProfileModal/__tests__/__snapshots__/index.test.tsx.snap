// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`EditProfileModal - Snapshot Tests should match snapshot when open 1`] = `
<div
  aria-hidden="true"
  data-aria-hidden="true"
/>
`;
